packages/app/src/request.ts

```diff
@@ -32,17 +37,64 @@ export const getProtocol = (req: Request, trust: Trust): Protocol => {
   return index !== -1 ? header.substring(0, index).trim() : header.trim()
 }
 
-export const getHostname = (req: Request, trust: Trust): string | undefined => {
-  let host: string = req.get('X-Forwarded-Host') as string
+const normalizeHostString = (host: string): string => decodeURIComponent(host).toLowerCase().normalize()
+
+const getAuthorityHeaderHostString = (req: Request): string | undefined => {
+  const authority = req.get(':authority')
+  if (Array.isArray(authority)) return undefined
+  if (!authority) return undefined
+
+  const index = authority.indexOf('@')
+  if (index === -1) return normalizeHostString(authority)
+  return normalizeHostString(authority.substring(index + 1))
+}
+
+const getForwardedHeaderHostString = (req: Request): string | undefined => {
+  const forwardedHost = req.get('x-forwarded-host')
+  if (Array.isArray(forwardedHost)) return undefined
+  if (!forwardedHost) return undefined
+
+  return normalizeHostString(forwardedHost)
+}
+
+const getDefaultHeaderHostString = (req: Request): string | undefined => {
+  const host = req.get('host')
+  if (!host) return undefined
```

Comment:
can be simplified with:

```suggestion
  if (!host || host.indexOf(',') !== -1) return undefined
```